cli/cmd/cluster/cluster.go

```diff
@@ -364,3 +366,90 @@ func (c controlplaneUpdater) ensureComponent(component, namespace string) error
 
 	return nil
 }
+
+// taintCertificates taints all certificate resources in existing Terraform
+// state. it will not taint the private keys of the CA so the public key gets
+// reused and the old CA cert can trust the new certificates.
+func (c *cluster) taintCertificates() error {
+	steps := []terraform.ExecutionStep{}
+
+	for _, t := range c.certificateResources() {
+		steps = append(steps, terraform.ExecutionStep{
+			Description: "taint certificate",
+			Args:        []string{"taint", t},
+		})
+	}
+
+	if err := c.terraformExecutor.Execute(steps...); err != nil {
+		return fmt.Errorf("tainting existing certificates: %w", err)
+	}
+
+	return nil
+}
+
+func (c *cluster) certificateResources() []string {
+	targets := []string{
+		// certificates
+		"tls_locally_signed_cert.admin",
+		"tls_locally_signed_cert.admission-webhook-server",
+		"tls_locally_signed_cert.aggregation-client[0]",
+		"tls_locally_signed_cert.apiserver",
+		"tls_locally_signed_cert.client",
+		"tls_locally_signed_cert.kubelet",
+		"tls_locally_signed_cert.peer",
+		"tls_locally_signed_cert.server",
+		"tls_self_signed_cert.aggregation-ca[0]",
+		"tls_self_signed_cert.etcd-ca",
+		"tls_self_signed_cert.kube-ca",
+		// Taint non-CA private keys as well, as those are safe to rotate.
+		"tls_private_key.admin",
+		"tls_private_key.admission-webhook-server",
+		"tls_private_key.aggregation-client[0]",
+		"tls_private_key.apiserver",
+		"tls_private_key.client",
+		"tls_private_key.kubelet",
+		"tls_private_key.peer",
+		"tls_private_key.server",
+	}
+
+	m := c.platform.Meta()
+
+	fullTargets := make([]string, 0, len(targets))
+
+	for _, target := range targets {
+		// The Tinkerbell platform doesn't set ClusterName and hardcodes the
+		// name of the controller module.
+		if m.Name == tinkerbell.Name {
+			fullTargets = append(fullTargets, fmt.Sprintf("module.controllers.module.bootkube.%s", target))
```

Comment:
We could perhaps pass controller module name via `Meta()`. Then we don't need `Name` and `ClusterName`?

Reply:
I don't like how `Meta()` is becoming more of a kitchen sink but I guess that's a bit better than the current implementation :+1:.